Support multiple players per user

evan-goode opened the pull request:
For https://github.com/unmojang/drasl/issues/75

- [x] Back-end functionality and refactoring
- [x] Disallow a user from claiming a Player Name the same as another user's Username
- [x] Disallow a user from claiming a Username the same as another user's Player Name
- [x] Send all connected players in `availableProfiles`
- [x] Support sending `selectedProfile` in /auth/refresh to ~~move Client to another Player of the same user~~ set `selectedProfile` for an unassigned client
- [x] Front-end changes 
  + [x] Allow admins to change an individual user's `MaxPlayerCount`
  + [x] Basic functionality
- [x] Write new tests
  - [x] availableProfiles, selectedProfile, refresh with selected profile
  - [x] Set max player count in admin view
  - [x] web create new player
  - [x] web register existing player, w/skin verification
  - [x] player name in use as another user's username
  - [x] username in use as another user's player name
  - [x] db migration
  - [x] db backup
- [x] Fix existing tests
- [x] Add create/update player API routes
- [x] Gracefully deprecate v1 API
- [x] Documentation
- [x] DB Migration
- [x] Automatic DB backup

IkyMax commented:
i tried this feature on my testing vps and i found two bugs, in the hypotetical case where the database is empty it'll say this

```
2024/12/24 00:29:32 Loading config from /etc/drasl/config.toml
2024/12/24 00:29:32 Started migration of database version 3 to 4.
2024/12/24 00:29:32 Backing up old database to /var/lib/drasl/drasl.3.2024-12-24T00-29-32Z.db
2024/12/24 00:29:32 Database backed up, proceeding.
2024/12/24 00:29:32 Error migrating database: empty slice found
```


and whenever i try to login with the main account (via yggdrasil REST API) it will only throw the available profiles with the client and access tokens, it won't throw the selected profile or the user section (if requested)

Request:

```
{
    "agent": {
        "name": "Minecraft",
        "version": 1
    },
    "username": "Test1",
    "password": "Test",
    "clientToken": "abcdef123456",
    "requestUser": true
}
```

Expected response:

```
{
    "accessToken": "accessToken",
    "clientToken": "abcdef123456",
    "selectedProfile": {
        "id": "0c947ccefad24cb4b62d62e3480d9860",
        "name": "Test1"
    },
    "availableProfiles": [{
        "id": "dd81cb374dae4f8889c9b5cb62c4be2f",
        "name": "Test1"
    }, {
        "id": "0c947ccefad24cb4b62d62e3480d9860",
        "name": "Test2"
    }],
    "user": {
        "id": "dd81cb374dae4f8889c9b5cb62c4be2f",
        "properties": [{
            "name": "preferredLanguage",
            "value": "en"
        }]
    }
}
```

Actual response:

```
{
    "accessToken": "accessToken",
    "clientToken": "abcdef123456",
    "availableProfiles": [{
        "id": "dd81cb374dae4f8889c9b5cb62c4be2f",
        "name": "Test1"
    }, {
        "id": "0c947ccefad24cb4b62d62e3480d9860",
        "name": "Test2"
    }]
}
```

evan-goode commented:
Fixed the empty slice error, it was a case of https://github.com/go-gorm/gorm/issues/4076, and added a regression test.

I also made `/authenticate` always honor `requestUser` and return the user ID instead of the selectedProfile ID. I _think_ this is the correct behavior according to authlib-injector: https://github-com.translate.goog/yushijinhun/authlib-injector/wiki/Yggdrasil-%E6%9C%8D%E5%8A%A1%E7%AB%AF%E6%8A%80%E6%9C%AF%E8%A7%84%E8%8C%83?_x_tr_sl=auto&_x_tr_tl=en&_x_tr_hl=en-US#%E7%94%A8%E6%88%B7%E4%BF%A1%E6%81%AF%E7%9A%84%E5%BA%8F%E5%88%97%E5%8C%96. I also updated the migration routine to generate new UUIDs for existing users so they are always distinct from any player UUID.

As far as the missing `selectedProfile`, IMO this is the more useful behavior. The authlib-injector spec isn't clear on disambiguating usernames/player names since it assumes you use email addresses as usernames. In my implementation, the username is matched first, then the player name. So if you have a user named `User1` with players `User1` and `User2`, selectedProfile will be empty and the launcher should prompt you which profile you want to use. But if user `Foo` owns players `Bar` and `Baz`, and you log in as `Bar` (no ambiguity), `selectedProfile` will be the `Bar` player.
